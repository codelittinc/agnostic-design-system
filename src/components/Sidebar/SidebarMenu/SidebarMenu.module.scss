@import '~@codelitt/ay-css-library/src/components/brand/colors';
@import '~@codelitt/ay-css-library/src/components/brand/grid';
@import '../Sidebar.module';

.menu {
  background: $ay-grey-8-color;
  height: 100vh;
  left: 0;
  outline: none;
  position: fixed;
  top: 0;
  transition: all 0.15s ease;
  width: $menu-width-desktop;
  z-index: 2;

  .mobile-header {
    display: none;
  }

  .list-items,
  .mobile-header {
    list-style: none;
    margin: 0;
    overflow: hidden;
    padding: 0;
  }

  .item-header {
    display: none;
  }
}

@media screen and (min-width: $view-width-tablet-min + 0.05rem) and (max-width: $view-width-desktop-min) {
  .menu {
    width: $menu-width-tablet;

    .list-items {
      overflow: visible;
    }
  }
}

@media screen and (max-width: $view-width-tablet-min) {
  .menu {
    height: $menu-height-phones;
    left: 0;
    outline: none;
    position: fixed;
    top: 0;
    transition: width 0.15s ease;
    width: 100%;

    .item-header {
      display: block;
    }

    .list-items {
      opacity: 0;
      overflow: visible;
      transition: visibility 0s, opacity 0.15s linear;
      visibility: hidden;

      &.open {
        opacity: 1;
        visibility: visible;
      }
    }

    .mobile-header {
      display: block;

      li {
        display: block;
      }
    }
  }
}
