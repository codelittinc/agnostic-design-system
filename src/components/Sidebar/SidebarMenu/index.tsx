import React, { useLayoutEffect, useState } from 'react';
import classnames from 'classnames';
import SidebarItem, {
  PropTypes as SidebarItemType,
} from '../SidebarItem/index';
import styles from './SidebarMenu.module.scss';

type PropTypes = {
  wrapperClassName?: string;
  items: Array<SidebarItemType>;
};

const SidebarMenu: React.FC<PropTypes> = (props: PropTypes) => {
  const [menuOpen, setMenuOpen] = useState(false);

  const closeMenu = () => {
    setMenuOpen(false);
  };

  useLayoutEffect(() => {
    if (menuOpen) {
      document.addEventListener('click', closeMenu);
    } else {
      document.removeEventListener('click', closeMenu);
    }

    return () => document.removeEventListener('click', closeMenu);
  }, [menuOpen]);

  const getActiveItemAsMenu = () => {
    let activeItem = props.items.find(item => item.isActive);

    if (!activeItem) {
      [activeItem] = props.items;
    }

    return {
      ...activeItem,
      onClick: () => setMenuOpen(!menuOpen),
    };
  };

  const getFixedItem = () => {
    return props.items.find(item => item.isFixed);
  };

  const renderSidebarItem = (
    item: SidebarItemType | undefined,
    key: string,
  ) => {
    if (!item) return null;

    return (
      <SidebarItem
        key={key}
        text={item.text}
        icon={item.icon}
        wrapperClassName={item.wrapperClassName}
        isActive={item.isActive}
        isFixed={item.isFixed}
        onClick={item.onClick}
      />
    );
  };

  return (
    <nav className={classnames(styles.menu, props.wrapperClassName)}>
      <ul className={classnames(styles['mobile-header'])}>
        {renderSidebarItem(getActiveItemAsMenu(), 'mobile-header-active')}
        {renderSidebarItem(getFixedItem(), 'mobile-header-fixed-item')}
      </ul>

      <ul
        className={classnames(styles['list-items'], {
          [styles.open]: menuOpen,
        })}
      >
        {props.items.map(item => renderSidebarItem(item, item.text))}
      </ul>
    </nav>
  );
};

export default SidebarMenu;
