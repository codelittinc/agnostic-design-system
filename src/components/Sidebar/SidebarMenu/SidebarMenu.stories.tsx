import * as React from 'react';
import SidebarMenu from './';
import { storiesOf } from '@storybook/react';

export default {
  title: 'SidebarMenu',
  component: SidebarMenu,
};

const stories = storiesOf('Components|Sidebar', module);

stories.add('SidebarMenu', () => <SidebarMenu items={[]} />, {
  order: 2,
});
