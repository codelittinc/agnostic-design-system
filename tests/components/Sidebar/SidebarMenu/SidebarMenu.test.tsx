import React from 'react';
import { render } from '@testing-library/react';
import SidebarMenu from '../../../../src/components/Sidebar/SidebarMenu';

describe('SidebarMenu', () => {
  const items = [
    {
      isActive: true,
      text: 'Menu text',
      onClick: () => {},
      icon: 'path/to/icon.svg',
    },
    {
      text: 'Menu text 2',
      onClick: () => {},
      icon: 'path/to/icon2.svg',
    },
  ];

  const getComponent = (wrapperClassName = '') => {
    return render(
      <SidebarMenu items={items} wrapperClassName={wrapperClassName} />,
    );
  };

  it('renders without crashing', () => {
    const sidebarMenu = getComponent();
    return expect(sidebarMenu!).toMatchSnapshot();
  });

  it('accepts a custom wrapper class', () => {
    const customClass = 'test-class-name';
    const { getByRole } = getComponent(customClass);
    const node = getByRole('navigation');

    expect(node).toHaveProperty('className', `menu ${customClass}`);
  });
});
