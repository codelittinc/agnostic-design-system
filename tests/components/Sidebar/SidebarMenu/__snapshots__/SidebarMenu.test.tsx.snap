// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarMenu renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="menu"
      >
        <ul
          class="mobile-header"
        >
          <li
            class="menu-item active"
          >
            <button
              class="menu-item-clickable"
              type="button"
            >
              <img
                alt="Menu text"
                class="menu-item-icon"
                src="path/to/icon.svg"
              />
              <span
                class="menu-item-text"
              >
                Menu text
              </span>
            </button>
          </li>
        </ul>
        <ul
          class="list-items"
        >
          <li
            class="menu-item active"
          >
            <button
              class="menu-item-clickable"
              type="button"
            >
              <img
                alt="Menu text"
                class="menu-item-icon"
                src="path/to/icon.svg"
              />
              <span
                class="menu-item-text"
              >
                Menu text
              </span>
            </button>
          </li>
          <li
            class="menu-item"
          >
            <button
              class="menu-item-clickable"
              type="button"
            >
              <img
                alt="Menu text 2"
                class="menu-item-icon"
                src="path/to/icon2.svg"
              />
              <span
                class="menu-item-text"
              >
                Menu text 2
              </span>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="menu"
    >
      <ul
        class="mobile-header"
      >
        <li
          class="menu-item active"
        >
          <button
            class="menu-item-clickable"
            type="button"
          >
            <img
              alt="Menu text"
              class="menu-item-icon"
              src="path/to/icon.svg"
            />
            <span
              class="menu-item-text"
            >
              Menu text
            </span>
          </button>
        </li>
      </ul>
      <ul
        class="list-items"
      >
        <li
          class="menu-item active"
        >
          <button
            class="menu-item-clickable"
            type="button"
          >
            <img
              alt="Menu text"
              class="menu-item-icon"
              src="path/to/icon.svg"
            />
            <span
              class="menu-item-text"
            >
              Menu text
            </span>
          </button>
        </li>
        <li
          class="menu-item"
        >
          <button
            class="menu-item-clickable"
            type="button"
          >
            <img
              alt="Menu text 2"
              class="menu-item-icon"
              src="path/to/icon2.svg"
            />
            <span
              class="menu-item-text"
            >
              Menu text 2
            </span>
          </button>
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
